### PR TITLE
enhance: GCS native support (GCS implemented using Google Cloud Storage libraries)

### PR DESCRIPTION
**Pull request by Rijin-N**
Native support for Google cloud storage using the Google Cloud Storage libraries. Authentication is performed using GCS service account credentials JSON.

Currently, Milvus supports Google Cloud Storage using S3-compatible APIs via the AWS SDK. This approach has the following limitations:

1. Overhead: Translating requests between S3-compatible APIs and GCS can introduce additional overhead.
2. Compatibility Limitations: Some features of the original S3 API may not fully translate or work as expected with GCS.

To address these limitations, This enhancement is needed.

Related Issue: #36212

### CODE REVIEW

**Comment by Rijin-N** [comment]
rerun go-sdk

**Comment by xiaofan-luan** [comment]
Thanks for the PR,
Quick question,  can you give us some potention compatibility problem of using S3 compatible SDK?

In our environment, we simply use S3 SDK to access GCS and it looks fine

File: configs/milvus.yaml [hunk 1]
```diff
@@ -133,11 +133,17 @@ minio:
...
   # You can use "aws" for other cloud provider supports S3 API with signature v4, e.g.: minio
   # You can use "gcp" for other cloud provider supports S3 API with signature v2
   # You can use "aliyun" for other cloud provider uses virtual host style bucket
+  # You can use "gcpnative" for the Google Cloud Platform provider. Uses service account credentials
+  # for authentication.
   # When useIAM enabled, only "aws", "gcp", "aliyun" is supported for now
   cloudProvider: aws
+  # The JSON content contains the gcs service account credentials.
+  # Used only for the "gcpnative" cloud provider.
+  gcpCredentialJSON:
```
Comment (shaoting-huang) [inline configs/milvus.yaml:146]: Please use cloud storage credential provider at [gcp.go](https://github.com/milvus-io/milvus/blob/526a672bae3b08c373af989a21e46d925c051a6b/internal/storage/gcp/gcp.go). Refer to [aliyun.go](https://github.com/milvus-io/milvus/blob/526a672bae3b08c373af989a21e46d925c051a6b/internal/storage/aliyun/aliyun.go#L23)

**Comment by shaoting-huang** [comment]
Thanks @Rijin-N for your contribution. Would you please provide the specific features GCS native has, which are needed for Milvus?

**Comment by Rijin-N** [comment]
> 2\. Compatibility Limitations: Some features of the original S3 API may not fully translate or work as expected with GCS.

@xiaofan-luan Thanks for the query, Some key compatibility limitations are listed below.

1. Object Versioning
**S3**: Amazon S3 supports object versioning, allowing multiple versions of the same object to be stored in a bucket. Each version is uniquely identified by a version ID.
**GCS:** GCS also supports object versioning, but the management and identification of object versions differ. In GCS, versions are identified using a generation number, not a version ID.
**Compatibility Limitation**: APIs that expect a version ID (like in S3) may not function the same way in GCS, requiring you to adjust how you work with versioning. Moreover, some advanced features related to versioning, such as multi-version delete markers, are handled differently.
2. ACLs (Access Control Lists)
**S3**: Amazon S3 has a robust ACL system, allowing for per-object and per-bucket permissions.
**GCS**: GCS supports both ACLs and IAM roles for permission management, but ACLs are discouraged in favor of IAM policies for broader, more fine-grained access control.
**Compatibility Limitation**: S3 ACL APIs may not directly map to GCS ACLs or IAM roles. IAM-based access control is different and more sophisticated than S3's ACLs
3. Bucket Policies
**S3:** S3 bucket policies are used to define access rules at the bucket level using JSON-based policy documents, allowing for fine-grained access control based on conditions like IP, time, or referrer.
**GCS:** GCS uses IAM policies, which are conceptually similar but differ in syntax, supported conditions, and features.
**Compatibility Limitation:** GCS does not fully support S3-style bucket policies. If you migrate bucket policies from S3 to GCS, you may need to reformat or rewrite them using IAM policies. 

Even though GCS works fine in milvus using AWS SDK,  there is always a possibility of future compatibility issues between Google Cloud Storage (GCS) and Amazon S3, especially as both services evolve and introduce new features. Below are some factors that could lead to future compatibility issues:

**New Features:** Both Amazon S3 and GCS continue to develop and release new features that may not immediately be supported by the S3 compatibility layer in GCS.
**API Changes:** If AWS updates the S3 API in a way that alters request formats, behaviors, or adds new parameters, it could cause compatibility problems for users relying on the GCS S3 API. GCS would need to continually update its compatibility layer to keep pace with these changes, which may not happen in real-time.

**Comment by xiaofan-luan** [comment]
@shaoting-huang 
please help on reviewing it.
@Rijin-N seems that there are some conflicts issue of the pr, please rebeas with master and rerun the ut

**Comment by Rijin-N** [comment]
> Thanks @Rijin-N for your contribution. Would you please provide the specific features GCS native has, which are needed for Milvus?

GCS native uses **GCS credentials JSON** file for authentication compared to AWS SDK access key and secret key methods.
Using GCS credentials JSON authentication is a more secure, flexible, and manageable approach, particularly for larger, production-grade deployments on Google Cloud. It also aligns well with best practices for cloud security and identity management.

Also Google Cloud Storage (GCS) SDK to access GCS provides following advantages over using the AWS SDK.

**1. Native Support and Optimized Performance:**

The GCS SDK is designed specifically for GCS, providing native support and being optimized for GCS’s architecture, features, and performance. This leads to better integration and potentially faster operations compared to using AWS SDK, which is not tailored to GCS.

**2. More Efficient Error Handling and Retry Logic:**

The GCS SDK implements error handling and retry logic tailored to GCS-specific error codes and network behavior, which ensures better handling of edge cases and recovery from transient issues.

**3. Authentication and Authorization:**

The GCS SDK natively supports Google Cloud IAM roles, OAuth tokens, and service accounts, making it easier to integrate with Google’s authentication mechanisms.

**4. Support and Documentation:**

The GCS SDK has comprehensive documentation and support directly from Google Cloud, making it easier to troubleshoot issues and ensure best practices when working with GCS.

File: configs/milvus.yaml (hunk 1 above)
Comment (Rijin-N) [reply configs/milvus.yaml:146]: GCS native uses GCS credentials JSON file for authentication using GCS go package. Refer file [gcp_native_object_storage.go](https://github.com/Rijin-N/milvus/blob/9782528a6369e7f7c1e0ef690c898b694b9f2864/internal/storage/gcp_native_object_storage.go#L59).

But [gcp.go](https://github.com/milvus-io/milvus/blob/526a672bae3b08c373af989a21e46d925c051a6b/internal/storage/gcp/gcp.go) and [aliyun.go](https://github.com/milvus-io/milvus/blob/526a672bae3b08c373af989a21e46d925c051a6b/internal/storage/aliyun/aliyun.go#L23) uses S3 compatible authetication with access key and secret key using minio go package. Both are different approches.

**Comment by Rijin-N** [comment]
Hi @xiaofan-luan I have rebased with master. Can you please help me to re-run the UTs.

File: scripts/core_build.sh [hunk 2]
```diff
@@ -101,6 +101,7 @@ USE_ASAN="OFF"
 USE_DYNAMIC_SIMD="ON"
 USE_OPENDAL="OFF"
 INDEX_ENGINE="KNOWHERE"
+: "${ENABLE_GCP_NATIVE:="ON"}"
```
Comment (shaoting-huang) [inline scripts/core_build.sh:104]: Since there is another dependency involved into Milvus. Let's set by default OFF

**Review by shaoting-huang — COMMENTED** [review]
Thanks for your contribution. I think let's turn off GCP Native by default and let user decide to use. Overall LGTM.

File: scripts/core_build.sh (hunk 2 above)
Comment (Rijin-N) [reply scripts/core_build.sh]: Done. Please check.

File: .env [hunk 3]
```diff
@@ -16,4 +16,5 @@ MINIO_ADDRESS=minio:9000
 PULSAR_ADDRESS=pulsar://pulsar:6650
 ETCD_ENDPOINTS=etcd:2379
 AZURITE_CONNECTION_STRING="DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://azurite:10000/devstoreaccount1;"
+ENABLE_GCP_NATIVE=ON
```
Comment (shaoting-huang) [inline .env:19]: sorry I didn't notice here. Please also change to OFF

Comment (Rijin-N) [reply .env:19]: I enable **ENABLE_GCP_NATIVE** to run the GCP-native unit tests (I believe this **.env** file is only used for unit testing purpose). If I turn it **OFF**, how will the GitHub workflow execute the GCP-native-specific unit tests?

**Comment by Rijin-N** [comment]
rerun go-sdk

**Comment by Rijin-N** [comment]
Hi, @shaoting-huang and @xiaofan-luan, Since the review comments have been updated and verified successfully, could you please assist with merging this PR? Let me know if any additional actions are required on my end. Thank you.

**Comment by xiaofan-luan** [comment]
> Hi, @shaoting-huang and @xiaofan-luan, Since the review comments have been updated and verified successfully, could you please assist with merging this PR? Let me know if any additional actions are required on my end. Thank you.

Merged! this should be released together with 2.5

**Comment by Rijin-N** [comment]
Thanks @shaoting-huang and @xiaofan-luan

**Pull request merged**